src/report.rs

```diff
@@ -71,6 +73,30 @@ pub struct Derived<P: Package, VS: VersionSet, M: Eq + Clone + Debug + Display>
 }
 
 impl<P: Package, VS: VersionSet, M: Eq + Clone + Debug + Display> DerivationTree<P, VS, M> {
+    /// Get all packages referred to in the derivation tree.
+    pub fn packages(&self) -> FxHashSet<&P> {
+        let mut packages = FxHashSet::default();
+        match self {
+            Self::External(external) => match external {
+                External::FromDependencyOf(p, _, p2, _) => {
+                    packages.insert(p);
+                    packages.insert(p2);
+                }
+                External::NoVersions(p, _)
+                | External::NotRoot(p, _)
+                | External::Custom(p, _, _) => {
+                    packages.insert(p);
+                }
+            },
+            Self::Derived(derived) => {
+                packages.extend(derived.terms.keys());
+                packages.extend(derived.cause1.packages().iter());
```

Comment:
Can we leave ourselves a breadcrumb comment here about how the recursive implementation is simple but what problems it might cause and how we could fix them?